File: PlantCatalog.cs
// PlantCatalog.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace PlantCatalog
{
    class Program
    {
        static void Main(string[] args)
        {
            var opts = ParseArgs(args);
            var catalog = new Catalog();
            if (opts.Add != null)
            {
                if (opts.Type == null || opts.Frequency == null)
                {
                    Console.Error.WriteLine("❌ Для добавления растения требуются --type и --frequency");
                    return;
                }
                catalog.AddPlant(opts.Add, opts.Type, opts.Frequency.Value, opts.Last);
            }
            else if (opts.List) catalog.ListPlants(opts.Filter);
            else if (opts.Water != null) catalog.WaterPlant(opts.Water);
            else if (opts.Remove != null) catalog.RemovePlant(opts.Remove);
            else if (opts.ExportJson != null) catalog.ExportJson(opts.ExportJson);
            else if (opts.ExportCsv != null) catalog.ExportCsv(opts.ExportCsv);
            else if (opts.ExportTxt != null) catalog.ExportTxt(opts.ExportTxt);
            else Console.WriteLine("Используйте --help для справки.");
        }

        static Options ParseArgs(string[] args)
        {
            var opts = new Options();
            for (int i = 0; i < args.Length; i++)
            {
                switch (args[i])
                {
                    case "--add": opts.Add = args[++i]; break;
                    case "--type": opts.Type = args[++i]; break;
                    case "--frequency": opts.Frequency = int.Parse(args[++i]); break;
                    case "--last": opts.Last = args[++i]; break;
                    case "--list": opts.List = true; break;
                    case "--filter": opts.Filter = args[++i]; break;
                    case "--water": opts.Water = args[++i]; break;
                    case "--remove": opts.Remove = args[++i]; break;
                    case "--export-json": opts.ExportJson = args[++i]; break;
                    case "--export-csv": opts.ExportCsv = args[++i]; break;
                    case "--export-txt": opts.ExportTxt = args[++i]; break;
                }
            }
            return opts;
        }

        class Options
        {
            public string Add { get; set; }
            public string Type { get; set; }
            public int? Frequency { get; set; }
            public string Last { get; set; }
            public bool List { get; set; }
            public string Filter { get; set; }
            public string Water { get; set; }
            public string Remove { get; set; }
            public string ExportJson { get; set; }
            public string ExportCsv { get; set; }
            public string ExportTxt { get; set; }
        }

        class Plant
        {
            public string Name { get; set; }
            public string Type { get; set; }
            public int Frequency { get; set; }
            public string LastWatered { get; set; }
        }

        class Catalog
        {
            private const string DataFile = "plants.json";
            private List<Plant> plants = new List<Plant>();

            public Catalog() => Load();

            private void Load()
            {
                try
                {
                    if (File.Exists(DataFile))
                    {
                        string json = File.ReadAllText(DataFile);
                        plants = JsonSerializer.Deserialize<List<Plant>>(json) ?? new List<Plant>();
                    }
                }
                catch { plants = new List<Plant>(); }
            }

            private void Save()
            {
                string json = JsonSerializer.Serialize(plants, new JsonSerializerOptions { WriteIndented = true });
                File.WriteAllText(DataFile, json);
            }

            public void AddPlant(string name, string type, int frequency, string last)
            {
                if (last == null) last = DateTime.UtcNow.ToString("yyyy-MM-dd");
                plants.Add(new Plant { Name = name, Type = type, Frequency = frequency, LastWatered = last });
                Save();
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine($"🌱 Растение '{name}' добавлено.");
                Console.ResetColor();
            }

            public void WaterPlant(string name)
            {
                var plant = plants.FirstOrDefault(p => p.Name.Equals(name, StringComparison.OrdinalIgnoreCase));
                if (plant == null)
                {
                    Console.ForegroundColor = ConsoleColor.Red;
                    Console.WriteLine($"❌ Растение '{name}' не найдено.");
                    Console.ResetColor();
                    return;
                }
                plant.LastWatered = DateTime.UtcNow.ToString("yyyy-MM-dd");
                Save();
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine($"💧 Растение '{name}' полито сегодня.");
                Console.ResetColor();
            }

            public void RemovePlant(string name)
            {
                var idx = plants.FindIndex(p => p.Name.Equals(name, StringComparison.OrdinalIgnoreCase));
                if (idx == -1)
                {
                    Console.ForegroundColor = ConsoleColor.Red;
                    Console.WriteLine($"❌ Растение '{name}' не найдено.");
                    Console.ResetColor();
                    return;
                }
                plants.RemoveAt(idx);
                Save();
                Console.ForegroundColor = ConsoleColor.Yellow;
                Console.WriteLine($"🗑️ Растение '{name}' удалено.");
                Console.ResetColor();
            }

            private (string status, ConsoleColor color) GetStatus(Plant p)
            {
                var last = DateTime.ParseExact(p.LastWatered, "yyyy-MM-dd", null);
                var daysSince = (DateTime.UtcNow - last).Days;
                if (daysSince <= p.Frequency) return ("✅", ConsoleColor.Green);
                if (daysSince <= p.Frequency * 1.5) return ("⚠️", ConsoleColor.Yellow);
                return ("🚨", ConsoleColor.Red);
            }

            public void ListPlants(string filter)
            {
                var list = plants;
                if (filter != null)
                {
                    list = plants.Where(p => p.Type.Equals(filter, StringComparison.OrdinalIgnoreCase)).ToList();
                    if (list.Count == 0)
                    {
                        Console.ForegroundColor = ConsoleColor.Yellow;
                        Console.WriteLine($"❌ Растения типа '{filter}' не найдены.");
                        Console.ResetColor();
                        return;
                    }
                }
                if (list.Count == 0)
                {
                    Console.ForegroundColor = ConsoleColor.Yellow;
                    Console.WriteLine("📭 Каталог пуст.");
                    Console.ResetColor();
                    return;
                }
                Console.ForegroundColor = ConsoleColor.Cyan;
                Console.WriteLine("🌿 Каталог растений:");
                Console.ResetColor();
                foreach (var p in list)
                {
                    var (status, color) = GetStatus(p);
                    var last = DateTime.ParseExact(p.LastWatered, "yyyy-MM-dd", null);
                    var next = last.AddDays(p.Frequency);
                    Console.ForegroundColor = color;
                    Console.WriteLine($"{status} {p.Name} ({p.Type}) - полив каждые {p.Frequency} дн., след. полив: {next:yyyy-MM-dd}");
                    Console.ResetColor();
                }
            }

            public void ExportJson(string filename)
            {
                string json = JsonSerializer.Serialize(plants, new JsonSerializerOptions { WriteIndented = true });
                File.WriteAllText(filename, json);
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine($"📄 Экспортировано в {filename} (JSON)");
                Console.ResetColor();
            }

            public void ExportCsv(string filename)
            {
                using var sw = new StreamWriter(filename);
                sw.WriteLine("name,type,frequency,last_watered");
                foreach (var p in plants)
                    sw.WriteLine($"{p.Name},{p.Type},{p.Frequency},{p.LastWatered}");
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine($"📄 Экспортировано в {filename} (CSV)");
                Console.ResetColor();
            }

            public void ExportTxt(string filename)
            {
                using var sw = new StreamWriter(filename);
                foreach (var p in plants)
                    sw.WriteLine($"{p.Name} | {p.Type} | {p.Frequency} дн. | последний полив: {p.LastWatered}");
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine($"📄 Экспортировано в {filename} (TXT)");
                Console.ResetColor();
            }
        }
    }
}

File: plant_catalog.js
#!/usr/bin/env node
// plant_catalog.js
const { program } = require('commander');
const fs = require('fs');
const chalk = require('chalk');

const DATA_FILE = 'plants.json';

class PlantCatalog {
    constructor() {
        this.plants = [];
        this.load();
    }

    load() {
        try {
            if (fs.existsSync(DATA_FILE)) {
                this.plants = JSON.parse(fs.readFileSync(DATA_FILE, 'utf8'));
            }
        } catch (e) {
            this.plants = [];
        }
    }

    save() {
        fs.writeFileSync(DATA_FILE, JSON.stringify(this.plants, null, 2));
    }

    addPlant(name, type, frequency, lastWatered) {
        if (!lastWatered) {
            lastWatered = new Date().toISOString().split('T')[0];
        }
        this.plants.push({ name, type, frequency, last_watered: lastWatered });
        this.save();
        console.log(chalk.green(`🌱 Растение '${name}' добавлено.`));
    }

    waterPlant(name) {
        const plant = this.plants.find(p => p.name.toLowerCase() === name.toLowerCase());
        if (!plant) {
            console.log(chalk.red(`❌ Растение '${name}' не найдено.`));
            return;
        }
        plant.last_watered = new Date().toISOString().split('T')[0];
        this.save();
        console.log(chalk.green(`💧 Растение '${name}' полито сегодня.`));
    }

    removePlant(name) {
        const idx = this.plants.findIndex(p => p.name.toLowerCase() === name.toLowerCase());
        if (idx === -1) {
            console.log(chalk.red(`❌ Растение '${name}' не найдено.`));
            return;
        }
        this.plants.splice(idx, 1);
        this.save();
        console.log(chalk.yellow(`🗑️ Растение '${name}' удалено.`));
    }

    getStatus(plant) {
        const last = new Date(plant.last_watered);
        const today = new Date();
        const daysSince = Math.floor((today - last) / (1000 * 60 * 60 * 24));
        const freq = plant.frequency;
        if (daysSince <= freq) return { status: '✅', color: chalk.green };
        if (daysSince <= freq * 1.5) return { status: '⚠️', color: chalk.yellow };
        return { status: '🚨', color: chalk.red };
    }

    listPlants(filterType) {
        let plants = this.plants;
        if (filterType) {
            plants = plants.filter(p => p.type.toLowerCase() === filterType.toLowerCase());
            if (plants.length === 0) {
                console.log(chalk.yellow(`❌ Растения типа '${filterType}' не найдены.`));
                return;
            }
        }
        if (plants.length === 0) {
            console.log(chalk.yellow('📭 Каталог пуст.'));
            return;
        }
        console.log(chalk.cyan('🌿 Каталог растений:'));
        for (const plant of plants) {
            const { status, color } = this.getStatus(plant);
            const last = new Date(plant.last_watered);
            const next = new Date(last);
            next.setDate(next.getDate() + plant.frequency);
            const nextStr = next.toISOString().split('T')[0];
            console.log(`${color(status)} ${plant.name} (${plant.type}) - полив каждые ${plant.frequency} дн., след. полив: ${nextStr}`);
        }
    }

    exportJson(filename) {
        fs.writeFileSync(filename, JSON.stringify(this.plants, null, 2));
        console.log(chalk.green(`📄 Экспортировано в ${filename} (JSON)`));
    }

    exportCsv(filename) {
        const header = 'name,type,frequency,last_watered\n';
        const rows = this.plants.map(p => `${p.name},${p.type},${p.frequency},${p.last_watered}`).join('\n');
        fs.writeFileSync(filename, header + rows);
        console.log(chalk.green(`📄 Экспортировано в ${filename} (CSV)`));
    }

    exportTxt(filename) {
        const lines = this.plants.map(p => `${p.name} | ${p.type} | ${p.frequency} дн. | последний полив: ${p.last_watered}`);
        fs.writeFileSync(filename, lines.join('\n'));
        console.log(chalk.green(`📄 Экспортировано в ${filename} (TXT)`));
    }
}

program
    .option('--add <name>', 'Добавить растение')
    .option('--type <type>', 'Тип растения')
    .option('--frequency <days>', 'Частота полива (дни)', parseInt)
    .option('--last <date>', 'Дата последнего полива (YYYY-MM-DD)')
    .option('--list', 'Показать все растения')
    .option('--filter <type>', 'Фильтр по типу')
    .option('--water <name>', 'Отметить растение политым')
    .option('--remove <name>', 'Удалить растение')
    .option('--export-json <file>', 'Экспорт в JSON')
    .option('--export-csv <file>', 'Экспорт в CSV')
    .option('--export-txt <file>', 'Экспорт в TXT')
    .parse(process.argv);

const opts = program.opts();
const catalog = new PlantCatalog();

if (opts.add) {
    if (!opts.type || !opts.frequency) {
        console.error(chalk.red('❌ Для добавления растения требуются --type и --frequency'));
        process.exit(1);
    }
    catalog.addPlant(opts.add, opts.type, opts.frequency, opts.last);
} else if (opts.list) {
    catalog.listPlants(opts.filter);
} else if (opts.water) {
    catalog.waterPlant(opts.water);
} else if (opts.remove) {
    catalog.removePlant(opts.remove);
} else if (opts.exportJson) {
    catalog.exportJson(opts.exportJson);
} else if (opts.exportCsv) {
    catalog.exportCsv(opts.exportCsv);
} else if (opts.exportTxt) {
    catalog.exportTxt(opts.exportTxt);
} else {
    program.help();
}
